content/blog/2025-04-17-blog.md

```diff
@@ -0,0 +1,57 @@
+
+# Highlights
+
+The authors begin with a discussion of downsides to typical implementations of SELF. First, the object prototype model, while it allows incredible flexibility, drastically increases the storage required for any SELF program. Secondly, message passing induces large penalties on runtime, as searching for a matching object slot can be very costly due to SELF’s multiple inheritance. The authors attack these twin costs of space and time with a set of mutually reinforcing techniques:
+- **Maps.** Maps provide implicit “hidden classes” for prototypes. Instead of storing slot names in every object, the VM factors layout and constant slots into a shared map. A Cartesian point shrinks from ten machine words to three, matching the compactness of class-based objects without reviving classes.
+
+- **Segregated heap for faster scanning.** Byte arrays live in their own region, so the garbage collector can sweep ordinary objects word-by-word without pausing to distinguish raw bytes from tagged pointers. This allowed SELF to scan memory at nearly double the rate of the fastest Smalltalk-80 implementation.
+
+- **Customized compilation.** At first send, SELF compiles a receiver-specific native method, locking in the receiver’s map and propagating that static knowledge through the body of the method.
+Message splitting and static type anticipation. When control-flow merges blur types, the compiler clones hot call-sites, keeps a “likely” branch where the receiver’s type is known, and leaves a slower fallback branch for rarer cases. The same mechanism lets the compiler anticipate that “+” or “<” will probably see integers and inlines those primitives aggressively.
+
+- **Zero-friction tooling.** Each compiled method records it's dependents; thus, a slot update invalidates only the affected machine code, leaving the rest of the system untouched. Rich metadata lets the debugger reconstruct inlined frames so developers see a clean, source-level stack even after heavy optimization.
+
+Taken together, these ideas deliver a dynamic language that is only four to five times slower than C on the same hardware — a leap forward at the time — and lay the conceptual groundwork for the hidden-class JITs that now power JavaScript, Ruby, and more.
+
+# Analysis
+
+In the over 30 years since this paper was published, there have been significant advances in the world of fast, efficient JIT compilers. Many of these developments stemmed directly from the groundwork laid in this paper, giving rise to the JIT compilers we know today - including the Java JIT compiler. After this paper was published, the described VM morphed into the Strongtalk engine which, after Sun bought Animorphic Systems in 1997, became the core of Java’s HotSpot VM. HotSpot still identifies “hot” call-sites, speculatively inlines them, and rolls back if reality diverges — precisely the speculative compilation playbook SELF introduced. 
```

Comment:
Maybe add a hyperlink to Strongtalk? This seems like something that readers might want to learn more about, and a link could help facilitate that.